### PR TITLE
Transaction: allow parameterizable buffer

### PR DESCRIPTION
**Pull request by dcousens**

### CODE REVIEW

**Comment by jprichardson** [comment]
utACK. Would you add a test?

**Comment by dcousens** [comment]
Going to put this in https://github.com/bitcoinjs/bitcoinjs-lib/pull/614,  so will add test today

**Comment by fanatid** [comment]
utACK

File: src/transaction.js [hunk 1]
```diff
@@ -304,6 +304,9 @@ Transaction.prototype.toBuffer = function () {
 
   writeUInt32(this.locktime)
 
+  // avoid slicing unless necessary
+  if (initialOffset !== undefined) return buffer.slice(initialOffset, offset)
```
Comment (dcousens) [inline src/transaction.js]: this will help in node environments,  but in the browser,  will this still result in a copy anyway?

**Comment by dcousens** [comment]
ping @jprichardson @fanatid ,  see comments

File: src/transaction.js (hunk 1 above)
Comment (fanatid) [inline src/transaction.js]: I don't understand, for what purpose are you using `slice` here?
In browser, `buffer` uses `subarray` for `slice`, i.e. if you change sliced array -- parent array also will be changed.

Comment (dcousens) [inline src/transaction.js]: Ok,  then not a concern!

Comment (dcousens) [inline src/transaction.js]: > I don't understand, for what purpose are you using slice here?

To encode the amount of data actually encoded into the `Buffer`.

**Comment by dcousens** [comment]
So @fanatid this is good to go then?

**Comment by fanatid** [comment]
I think yes.
ACK

**Comment by dcousens** [comment]
ping @jprichardson

**Comment by jprichardson** [comment]
It LGTM, but I should have originally asked the most important question... why?

**Comment by dcousens** [comment]
@jprichardson performance!  Noticed some fun things with lots of large transactions,  saved myself quite a few `Buffer` allocations using this method.

**Pull request merged**